doc/sphinx/source/libCEEDdev.md

```diff
@@ -1,98 +1,109 @@
+The `/cpu/self/avx/*` and `/cpu/self/xsmm/*` backends delegate to the corresponding `/cpu/self/opt/*` backends.
+These backends update the `CeedTensorContract` objects.
 
-Header files should be listed in alphabetical order, with installed headers preceding local headers and `ceed` headers being listed first.
-The `ceed-f64.h` and `ceed-f32.h` headers should only be included in `ceed.h`.
+The `/cpu/self/memcheck/*` backends are delegate to the `/cpu/self/ref/*` backends.
+These backends replace many of the implementations with methods that include more verification checks and a memory management model that more closely matches the memory management for GPU backends.
 
-```c
-#include <ceed.h>
-#include <ceed/backend.h>
-#include <stdbool.h>
-#include <string.h>
-#include "ceed-avx.h"
-```
+### GPU Backends
 
-## Shape
+The CUDA, HIP, and SYCL backend families all follow the same basic design.
 
-Backends often manipulate tensors of dimension greater than 2.
-It is awkward to pass fully-specified multi-dimensional arrays using C99 and certain operations will flatten/reshape the tensors for computational convenience.
-We frequently use comments to document shapes using a lexicographic ordering.
-For example, the comment
+The `/gpu/*/ref` backends provide basic functionality.
+In these backends, the operator is applied in multiple separate kernel launches, following the libCEED operator decomposition, where first {ref}`CeedElemRestriction` kernels map from the L-vectors to E-vectors, then {ref}`CeedBasis` kernels map from the E-vectors to Q-vectors, then the {ref}`CeedQFunction` kernel provides the action of the user quadrature point function, and the transpose {ref}`CeedBasis` and {ref}`CeedElemRestriction` kernels are applied.
+These kernels apply to all points across all elements in order to maximize the amount of work each kernel launch has.
 
-```c
-// u has shape [dim, num_comp, Q, num_elem]
-```
+The `/gpu/*/shared` backends delegate to the corresponding `/gpu/*/ref` backends.
+These backends use shared memory to reduce the amount of memory required for the {ref}`CeedBasis` kernels.
 
-means that it can be traversed as
+The `/gpu/*/gen` backends delegate to the corresponding `/gpu/*/shared` backends.
+These backends write a single comprehensive kernel to apply the action of the {ref}`CeedOperator`, significantly improving performance by eliminating intermediate data structures and reducing the total number of kernel launches required.
 
-```c
-for (d=0; d<dim; d++)
-  for (c=0; c<num_comp; c++)
-    for (q=0; q<Q; q++)
-      for (e=0; e<num_elem; e++)
-        u[((d*num_comp + c)*Q + q)*num_elem + e] = ...
-```
+The `/gpu/*/magma` backends delegate to the corresponding `/gpu/*/ref` backends.
+These backends provide better performance for {ref}`CeedBasis` kernels but do not have the improvements from the `/gpu/*/gen` backends for {ref}`CeedOperator`.
```

Comment:
If magma only helps with the basis, could/should it delegate to `/gpu/*/shared` instead? If not, why not?

Reply:
It should not - shared replaces on the CeedBasis, and that's also what MAGMA replaces, so it would be an extra layer of indirection that isn't needed

Reply:
Understood!